File: symmetric/example4/encrypt.js
const fs = require('fs');
const crypto = require('crypto');

// Считывание симметричного ключа из файла
const key = fs.readFileSync('symmetric_key.bin');

// Создание случайного вектора инициализации (IV) длиной 12 байт
const iv = crypto.randomBytes(12);

// Сообщение, которое нужно зашифровать
const message = 'Это секретное сообщение, которое мы хотим зашифровать.';

// Создание объекта cipher для шифрования симметричным ключом AES
const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);

// Шифрование сообщения
const encryptedMessage = Buffer.concat([
  cipher.update(message, 'utf8'),
  cipher.final(),
]);

// Получение аутентификационного тега
const authTag = cipher.getAuthTag();

// Комбинирование IV, зашифрованных данных и аутентификационного тега
const encryptedMessageWithAuthTag = Buffer.concat([
  iv,
  encryptedMessage,
  authTag,
]);

// Запись зашифрованного сообщения в файл
fs.writeFileSync('encrypted_message.bin', encryptedMessageWithAuthTag);

console.log('Сообщение успешно зашифровано и записано в файл.');

File: symmetric/example3/encrypt.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"fmt"
	"io"
	"os"
)

func main() {
	// Считывание симметричного ключа из файла
	key, err := readKeyFromFile("symmetric_key.bin")
	if err != nil {
		fmt.Println("Ошибка при считывании ключа из файла:", err)
		return
	}

	// Сообщение, которое нужно зашифровать
	message := "Это секретное сообщение, которое мы хотим зашифровать."

	// Шифрование сообщения
	encryptedMessage, err := encryptMessage(key, message)
	if err != nil {
		fmt.Println("Ошибка при шифровании сообщения:", err)
		return
	}

	fmt.Println("Зашифрованное сообщение:")
	fmt.Printf("%x\n", encryptedMessage)

	// Запись зашифрованного сообщения в файл
	err = writeToFile("encrypted_message.bin", encryptedMessage)
	if err != nil {
		fmt.Println("Ошибка при записи зашифрованного сообщения в файл:", err)
		return
	}

	fmt.Println("Зашифрованное сообщение успешно записано в файл.")
}

// Считывание ключа из файла
func readKeyFromFile(filename string) ([]byte, error) {
	key, err := os.ReadFile(filename)
	return key, err
}

// Шифрование сообщения с использованием симметричного ключа
func encryptMessage(key []byte, message string) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, err
	}

	// Шифрование сообщения
	encryptedMessage := gcm.Seal(nonce, nonce, []byte(message), nil)

	return encryptedMessage, nil
}

// Запись ключа в файл
func writeToFile(filename string, data []byte) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	_, err = file.Write(data)
	return err
}